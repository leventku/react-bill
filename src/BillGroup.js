import React, { Component } from 'react';

export default class BillGroup extends Component {
  constructor(props) {
    super(props);
    const minVisible = 3;
    this.state = {
      minVisible,
      maxVisible: props.items.length,
      visibleRows: minVisible
    }
  }

  toggleShowAll = () => {
    this.setState((prevState) => {
      if (prevState.visibleRows === prevState.minVisible) {
        return {visibleRows: prevState.maxVisible}
      }
      else {
        return {visibleRows: prevState.minVisible}
      }
    })
  }

  render() {
    return (
      <div className="bill-group">
        <h2>{this.props.title}: Â£{this.props.total.toFixed(2)}</h2>
        <div className="bill-list-container">
          <ul className="list-group">
            {
              this.props.items.slice(0, this.state.visibleRows).map(this.props.itemRenderer)
            }
          </ul>
          <button onClick={this.toggleShowAll} style={{display: this.state.maxVisible > this.state.minVisible ? 'inline-block': 'none'}}>
            {this.state.visibleRows === this.state.minVisible ? 'Show All': 'Show Less'}
          </button>
        </div>
      </div>
    )
  }
}