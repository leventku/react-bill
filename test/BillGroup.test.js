import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import BillGroup from '../src/BillGroup';

function setupComponent() {
  const props = {
    title: 'Some title',
    items: [1, 2, 3, 4, 5],
    itemRenderer: jest.fn(),
    total: 123,
  };

  const wrapper = shallow(<BillGroup {...props} />);

  return {
    props,
    wrapper,
  };
}

describe('BillGroup component', () => {
  it('should render', () => {
    const { wrapper } = setupComponent();
    expect(wrapper.exists()).toBe(true);
  });

  it('should render its items', () => {
    const { wrapper } = setupComponent();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call itemRenderer function 3 times initially', () => {
    const { props } = setupComponent();
    expect(props.itemRenderer.mock.instances.length).toBe(3);
  });

  it('should call itemRenderer function 5 times after show all button is clicked', () => {
    const { props, wrapper } = setupComponent();
    wrapper.find('button').simulate('click');
    expect(props.itemRenderer.mock.instances.length).toBe(8);
    wrapper.find('button').simulate('click');
    expect(props.itemRenderer.mock.instances.length).toBe(11);
  });
});
