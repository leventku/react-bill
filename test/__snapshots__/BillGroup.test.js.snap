// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillGroup component should render its items 1`] = `
<div
  className="bill-group"
>
  <h2>
    Some title
     Â£
    123.00
  </h2>
  <div
    className="bill-list-container"
  >
    <ul
      className="list-group"
    />
    <button
      onClick={[Function]}
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      Show All
    </button>
  </div>
</div>
`;
